<launch>
  <arg name="FIXED_FRAME" default="/map" />
  <arg name="USE_ASSEMBLER_FILTER" default="true" />
  <!-- laser assembler for recognizing environment -->
  <node pkg="laser_assembler" type="laser_scan_assembler"
        name="laser_scan_assembler_node"
        output="screen" >
    <remap from="scan" to="/multisense_sl/laser/scan"/>
    <remap from="cloud_filtered" to="/tilt_scan_shadow2_filtered"/>
    <param name="fixed_frame" type="string" value="$(arg FIXED_FRAME)"/>
    <rosparam command="load" file="$(find hrpsys_gazebo_atlas)/config/atlas_laser_filter.yaml" if="$(arg USE_ASSEMBLER_FILTER)"/>
  </node>
  <node pkg="hrpsys_gazebo_atlas" type="atlas_laser_snapshotter"
        name="atlas_laser_snapshotter"
        output="screen" >
    <remap from="joint_states" to="/multisense_sl/joint_states" />
    <remap from="full_cloud2" to="/laser/full_cloud2_raw" />
    <!-- publish /full_cloud2 -->
  </node>
  <node pkg="robot_self_filter"
        type="self_filter" clear_params="true"
        name="robot_shadow_self_filter" respawn="true" output="screen">
    <!-- The topic for the input cloud -->
    <remap from="cloud_in" to="/laser/full_cloud2_raw" />
    <!-- The topic for the output cloud -->
    <remap from="cloud_out" to="/laser/full_cloud2_filtered" />
    <!-- The frame of the sensor used to obtain the data to be
	 filtered; This parameter is optional. If it is not specified,
         shadow points will be considered outside -->
    <!--param name="sensor_frame" type="string" value="laser_tilt_link" /-->
    <!-- Minimum distance to sensor (for point not to be considered inside)-->
    <param name="subsample_value" type="double" value="0.0"/>
    <!-- The names of the links the sensor can see -->
    <rosparam command="load" file="$(find hrpsys_gazebo_atlas)/config/tilt_self_filter.yaml" />
  </node>

  <!-- laser to point cloud for moveit -->
  <node pkg="laser_filters" type="scan_to_cloud_filter_chain"
	name="atlas_scan_to_cloud"
	output="screen" clear_params="true">
    <remap from="scan" to="/multisense_sl/laser/scan" />
    <remap from="cloud_filtered" to="/laser/tilt_cloud_filtered" />
    <!--param name="target_frame" value="/pelvis" /-->
    <!--param name="target_frame" value="/head" /-->
    <param name="target_frame" value="/head_hokuyo_frame" />
    <rosparam command="load" file="$(find hrpsys_gazebo_atlas)/config/atlas_scan_to_cloud.yaml" />
  </node>

  <!-- pusblish dummy cloud for moveit -->
  <node name="empty_cloud_publisher" pkg="jsk_pr2_startup"
	output="screen"
        type="publish_empty_cloud.py">
    <param name="frame_id" value="/head_hokuyo_frame" />
    <param name="max_range" value="5.0" />
    <param name="rate" value="3" />
    <param name="rotate_points" value="true" />
  </node>
  <node name="atlas_laser_mux" pkg="topic_tools" type="mux"
        args="/laser/moveit_input_cloud2 /laser/tilt_cloud_filtered /empty_cloud ">
    <remap from="mux" to="atlas_laser_mux"/>
  </node>

  <!-- start tilting -->
  <node name="start_tilt_laser_node"
        type="start-laser.py" pkg="hrpsys_gazebo_atlas"
        output="screen" args="1.1424"/> <!-- 2pi / 5.5 sec -->
  <node name="start_tilt_laser_node_dyn"
	type="dynparam" pkg="dynamic_reconfigure"
	args="set /multisense_sl motor_speed 1.1424"
	output="screen" />
</launch>
